src/client/dfs/dfs_internal.c

```diff
@@ -7,8 +7,248 @@
 #include <daos/common.h>
 #include "dfs_internal.h"
 
+/** protect against concurrent dfs_init/fini calls */
+static pthread_mutex_t	module_lock = PTHREAD_MUTEX_INITIALIZER;
+
+/** refcount on how many times dfs_init has been called */
+static int		module_initialized;
+
+/** hashtable for pool open handles */
+static struct d_hash_table *poh_hash;
+
+/** hashtable for container open handles */
+static struct d_hash_table *coh_hash;
+
+static inline struct dfs_mnt_hdls*
+hdl_obj(d_list_t *rlink)
+{
+	return container_of(rlink, struct dfs_mnt_hdls, entry);
+}
+
+static bool
+key_cmp(struct d_hash_table *htable, d_list_t *rlink, const void *key, unsigned int ksize)
+{
+	struct dfs_mnt_hdls *hdl = hdl_obj(rlink);
+
+	return (strcmp(hdl->value, key) == 0);
+}
+
+static void
+rec_addref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	hdl_obj(rlink)->ref++;
+}
+
+static bool
+rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dfs_mnt_hdls *hdl = hdl_obj(rlink);
+
+	assert(hdl->ref > 0);
+	hdl->ref--;
+	return (hdl->ref == 0);
+}
+
+static void
+rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dfs_mnt_hdls *hdl = hdl_obj(rlink);
+
+	assert(d_hash_rec_unlinked(&hdl->entry));
+	assert(hdl->ref == 0);
+
+	if (hdl->type == DFS_H_POOL)
+		daos_pool_disconnect(hdl->handle, NULL);
+	else if (hdl->type == DFS_H_CONT)
+		daos_cont_close(hdl->handle, NULL);
+	else
+		D_ASSERT(0);
+	D_FREE(hdl);
+}
+
+static uint32_t
+rec_hash(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dfs_mnt_hdls *hdl = hdl_obj(rlink);
+
+	return d_hash_string_u32(hdl->value, strlen(hdl->value));
+}
+
+static d_hash_table_ops_t hdl_hash_ops = {
+	.hop_key_cmp = key_cmp,
+	.hop_rec_addref = rec_addref,
+	.hop_rec_decref = rec_decref,
+	.hop_rec_free = rec_free,
+	.hop_rec_hash = rec_hash
+};
+
+int
+dfs_is_init()
+{
+	if (module_initialized > 0)
+		return true;
+	return false;
+}
+
+int
+dfs_init()
+{
+	int	rc;
+
+	D_MUTEX_LOCK(&module_lock);
+	if (module_initialized > 0) {
+		/** already initialized, report success */
+		module_initialized++;
+		D_GOTO(unlock, rc = 0);
+	}
+
+	rc = daos_init_locked();
+	if (rc)
+		return daos_der2errno(rc);
+
+	rc = d_hash_table_create(D_HASH_FT_EPHEMERAL | D_HASH_FT_LRU, 4, NULL, &hdl_hash_ops,
+				 &poh_hash);
+	if (rc) {
+		D_ERROR("Failed to init pool handle hash "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out_fini, rc = daos_der2errno(rc));
+	}
+
+	rc = d_hash_table_create(D_HASH_FT_EPHEMERAL | D_HASH_FT_LRU, 4, NULL, &hdl_hash_ops,
+				 &coh_hash);
+	if (rc) {
+		D_ERROR("Failed to init container handle hash "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out_poh, rc = daos_der2errno(rc));
+	}
+
+	module_initialized++;
+	D_GOTO(unlock, rc = 0);
+
+out_poh:
+	d_hash_table_destroy(poh_hash, false);
+out_fini:
+	daos_fini();
```

Comment:
If we can use `daos_init_locked` above, then could we use `daos_fini_locked` here? (but also see my comment above)